### PR TITLE
Absolute arrow tail on annotations

### PR DESCRIPTION
**Pull request by mdfederici**
Adding the ability to specify the tail of an annotation arrow in absolute point in grid terms rather than relative pixel offset terms.

This allows the creation of trendlines which will scale correctly when the plot is zoomed.

### CODE REVIEW

**Comment by mdfederici** [comment]
When i run the tests locally, two of them fail on my machine although they fail for me on a fresh pull from plotly master so I do not believe my code caused this.

I have not added any tests for my implementation. I would like to but I don't really know where the appropriate place is for them (and should they be jasmine vs image based). I sort of was expecting to find /test/jasmine/tests/annotation_test.js but didn't find one. Some other jasmine tests seem to specify annotations but aren't specifically attempting to test annotations. Any guidance on this would be appreciated so that I can add some appropriate tests.

**Comment by mdtusz** [comment]
Great looking PR at first glance - I'll have a closer look through sometime over the weekend. 

Regarding tests, it would be useful to have both an image and jasmine test. Feel free to add a new file `annotations_test.js` and I'm sure we will also add some as issues arise as well. You can just add one of these absolute arrows to the existing `annotations.json` image test mock though.

**Comment by mdfederici** [comment]
Good catch on the editing bit - i hadn't considered that. The behavior was such that the tail would stay fixed and the arrow head would move. When thinking about how people are likely to want to 'edit' a trendline, it seems to me to be more likely that they'll want to move it out of the way a bit or maybe compare it to some other part of the data and as such, the slope should remain fixed. I've amended my commit to reflect this. I didn't add code to do this in the similar code for dragging the text. My reasoning is that I believe the reason to drag the text is to move it out of the way but for a trendline (given the above behavior of maintaining slope) this isn't really a use case. In fact if people mostly are using this functionality for a trendline, its likely that they won't attach text to the arrow at all. Still, it would be nice if dragging the text when absoluteArrowTail=true was equivalent to dragging the arrow itself but i think that writing the code to do that will either a) be noisy relative to the value or b) require a large refactor to avoid code duplication/noise.

Thoughts?

I also added tests.

Also is there a standard for js version - or more specifically can i use es6 constructs such as 'let'? I'm guessing no because i don't see usage anywhere else and probably you want some older browser compatibility.

**Comment by mdfederici** [comment]
Also can you let me know the publishing schedule for plotly.js? I'm wondering how long after we get this accepted to expect to see the changes on [npm](https://www.npmjs.com/package/plotly.js).

**Comment by mdtusz** [comment]
We do not use es6 on plotly.js unfortunately, so no `let`.

Editing arrows is mainly used in the www.plot.ly webapp, so we need to maintain consistency in behaviour for it - for arrows such as this, I do not anticipate that they will _only_ be used for trendlines, so an ideal behaviour would be when dragging the head or tail, changing either the `x/ax` or `y/ay` the same as is done now, but setting the value appropriate to the `absoluteArrowTail` setting. 

Our release schedule is not exactly regular, but when we have what we consider a "good" state with enough substantial changes, we will make a release - typically this is every week or two. Our last release was 11 days ago so if this is wrapped up in the next day or two, we'll likely have it out either tomorrow or Wednesday, or Monday (we don't like to release on Fridays).

File: src/components/annotations/attributes.js [hunk 1]
```diff
@@ -131,14 +131,31 @@ module.exports = {
         role: 'style',
         description: 'Sets the width (in px) of annotation arrow.'
     },
+    absoluteArrowTail: {
```
Comment (mdtusz) [inline src/components/annotations/attributes.js]: Although not convention for javascript, we use all lowercased names for attributes. I'd suggest changing the name on this to just `absolutetail` though.

**Comment by mdtusz** [comment]
Also, could you make a pull request on the plotly/plotly.js repo please? That way, it's easier for us to pull down the branch to test and keep track of things.

File: src/components/annotations/attributes.js (hunk 1 above)
Comment (cldougl) [inline src/components/annotations/attributes.js]: +1 for lowercased, and imo `absolutetail` is descriptive enough. 
(nice descriptive description!)

**Comment by mdfederici** [comment]
ok so i went through with the text and editing and so forth and i think i have it all working as desired. I'm getting really wonky/inconsistent behavior with the tests but i'm getting those intermittent test failures without my changes so I don't think i'm causing them. Please let me know if you can confirm if I am or am not causing any tests to fail.

I've created a pull request to plotly/plotly.js which is logically equivalent to this one so we can move over [there](https://github.com/plotly/plotly.js/pull/610).
@mdtusz

**Pull request closed without merging**